teerex/src/lib.rs

```diff
@@ -105,27 +105,45 @@ pub mod pallet {
 		/// Verifying RA report failed.
 		RemoteAttestationVerificationFailed,
 		/// IAS remote attestation is too old
-		RemoteAttestationTooOld,
+		RemoteAttestationIsTooOld,
 		/// The enclave cannot attest, because its building mode is not allowed.
-		SgxModeNotAllowed,
+		SgxModeIsNotAllowed,
 		/// The enclave is not registered.
 		EnclaveIsNotRegistered,
 		/// The worker url is too long.
-		EnclaveUrlTooLong,
+		EnclaveUrlIsTooLong,
 		/// The Remote Attestation proof is too long.
-		RaProofTooLong,
+		RaProofIsTooLong,
 		/// No enclave is registered.
 		EmptyEnclaveRegistry,
 		/// The provided collateral data is invalid
-		CollateralInvalid,
+		CollateralIsInvalid,
 		/// It is not allowed to unregister enclaves with recent activity
-		UnregisterActiveEnclaveNotAllowed,
+		UnregisterActiveEnclaveIsNotAllowed,
 		/// skipping attestation not allowed by configuration
-		SkippingAttestationNotAllowed,
+		SkippingAttestationIsNotAllowed,
 		/// No TCB info could be found onchain for the examinee's fmspc
 		MissingTcbInfoForFmspc,
 		/// Either the enclave TCB has outdated status or the onchain TCB collateral is outdated
-		TcbInfoOutdated,
+		TcbInfoIsOutdated,
+		/// The number of certificates should be >=2
+		CertificateChainIsTooShort,
+
+		SgxCaVerificationFailed,
+		SgxRsaSignatureIsInvalid,
+		/// An error originating in the sgx_verify crate
+		OtherSgxVerifyError(sgx_verify::Error),
+	}
+
+	impl<T> From<sgx_verify::Error> for Error<T> {
+		fn from(e: sgx_verify::Error) -> Self {
+			use sgx_verify::Error as Theirs;
+			match e {
+				Theirs::CaVerificationFailed => Self::SgxCaVerificationFailed,
+				Theirs::RsaSignatureIsInvalid => Self::SgxRsaSignatureIsInvalid,
+				_ => Self::OtherSgxVerifyError(e),
+			}
+		}
```

Comment:
tried to impl your suggestion @clangenb  but the compiler can't find this implementation of From:
```
the trait `From<sgx_verify::Error>` is not implemented for `sp_runtime::DispatchError`
    |
    = note: the question mark operation (`?`) implicitly performs a conversion on the error value using the `From` trait
    = help: the following other types implement trait `From<T>`:
              <sp_runtime::DispatchError as From<&'static str>>
              <sp_runtime::DispatchError as From<ArithmeticError>>
              <sp_runtime::DispatchError as From<TokenError>>
              <sp_runtime::DispatchError as From<TransactionalError>>
              <sp_runtime::DispatchError as From<frame_system::Error<T>>>
              <sp_runtime::DispatchError as From<hidden_include::error::BadOrigin>>
              <sp_runtime::DispatchError as From<hidden_include::error::LookupError>>
              <sp_runtime::DispatchError as From<pallet::Error<T>>>
              <sp_runtime::DispatchError as From<pallet_balances::Error<T, I>>>
    = note: required for `sgx_verify::Error` to implement `Into<sp_runtime::DispatchError>`
    = note: required for `DispatchErrorWithPostInfo<PostDispatchInfo>` to implement `From<sgx_verify::Error>`
    = note: required for `Result<([u8; 6], teerex_primitives::SgxTcbInfoOnChain), DispatchErrorWithPostInfo<PostDispatchInfo>>` to implement `FromResidual<Result<Infallible, sgx_verify::Error>>`

```

Reply:
expanded the macros but couldn't find the cause

Reply:
this shouldn'r be necessary, but it works if I'm very explicit:
`.map_err(|e| Error::<T>::from(e))?`
not as nice as a simple `?`, but I can live with this

Reply:
I believe that the problem is that we are doing a two layered conversion and I think the compiler can't handle that: `sgx_verify::Error -> Error<T> -> sp_runtime::DispatchError`

I am quite positive that this doesn't have to do anything with the macro magic, but the rust compiler simply can't do that.

Reply:
The only way to fix this would be to introduce a wrapper function, but I don't think it is worth it:

```rust
fn verify_ias_report(..) -> Error<T> {
     sgx_verify::sgx_verify_ias_report(...)?:
}

// somewhere in the pallet:
Self::verify_ias_report()?;
```

Reply:
the readability is better as-is IMO